pkg/threescale/backend/backend.go

```diff
@@ -253,11 +257,222 @@ func (b *Backend) localReport(cacheKey string, metrics api.Metrics) {
+
+		// calculate the deltas and report remotely
+		deltas := appClone.calculateDeltas()
+		// store the deltas for further use
+		apps[i].deltas = deltas
+		req := threescale.Request{
+			Auth:    appClone.auth,
+			Service: cachedApp.serviceID,
+			Transactions: []api.Transaction{
+				{
+					Metrics: deltas,
+					Params:  app.params,
+				},
+			},
+			Extensions: api.Extensions{
+				api.FlatUsageExtension: "1",
+			},
+		}
+		_, err := b.remoteReport(req)
+		if err != nil {
+			apps[i].reportingErr = true
+		}
+	}
+}
+
+// handleFlushCacheUpdate updates a cached entry (taking a write lock) using the prior flushing steps as a decision tree.
+// A description of the terminology used below is as follows:
+//     snapshot_hits - the value in local state when the snapshot is taken
+//     new_auth_hits - the current value for a metric as fetched from 3scale after successful authorization
+//     to_report_hits - the deltas that are calculated prior to reporting = current local state - last known 3scale state
+//     actually_reported_hits - is equal 0 (if report fails) || is equal to to_report_hits (if report succeeded)
+// We need to handle the following cases ("hits" is used as an example metric name for simplification):
+// 1. reporting error and authorization error -> nothing to do, no change of state
+// 2. reporting error and authorization success -> hits += new_auth_hits - (snapshot_hits - to_report_hits ) - actually_reported_hits
+// 3. report success and authorization error -> new remote state = current remote state + reported deltas. No change to local counters.
+// 4. report success and authorization success -> hits += new_auth_hits - (snapshot_hits - to_report_hits ) - actually_reported_hits
+// To summarise, if we have a reporting error, we will not amend our local representation of the remote state.
+// If we have an authorization error we will not amend our local state.
+func (b *Backend) handleFlushCacheUpdate(apps []cachedApp) {
+	for _, cachedApp := range apps {
+		// handle case (1)
+		if cachedApp.reportingErr && cachedApp.authErr {
+			// nothing to do here only continue on
+			// todo add logging
+			continue
+		}
+
+		//now that we have handled the dual error case, we know we need to
+		// read from the cache and modify the entry so we take a write lock
+		app, ok := b.cache.Get(cachedApp.cacheKey)
+		if !ok {
+			continue
+		}
+
+		// handle case (3)
+		if !cachedApp.reportingErr && cachedApp.authErr {
+			app.Lock()
+			// This is necessary when, during flushing, we have reported correctly but failed to authorize
+			// we dont have an updated state from 3scale so in order to avoid re-reporting what we have already
+			// successfully reported. We can achieve this safely by modifying our last know local
+			// representation of the remote state with the deltas we know that 3scale has processed.
+			app.addDeltasToRemoteState(cachedApp.deltas)
+			app.UnlimitedCounter = make(UnlimitedCounter)
```

Comment:
Could this potentially be overwritten here? What if a new request just reported some unknown metrics to this app after snapshotting and before taking the lock?